roachtest: sqlsmith/setup=empty/setting=no-ddl failed

roachtest.sqlsmith/setup=empty/setting=no-ddl [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5902767?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/5902767?buildTab=artifacts#/sqlsmith/setup=empty/setting=no-ddl) on master @ [1129fbc650fe3a037b03aea1e5f1d8078618cb1c](https://github.com/cockroachdb/cockroach/commits/1129fbc650fe3a037b03aea1e5f1d8078618cb1c):


```
test artifacts and logs in: /artifacts/sqlsmith/setup=empty/setting=no-ddl/run_1
	sqlsmith.go:265,sqlsmith.go:325,test_runner.go:896: error: pq: internal error: unexpected error from the vectorized engine: interface conversion: coldata.Column is *coldataext.datumVec, not *coldata.Bytes
		stmt:
		SELECT
			(-3781271420333609070):::INT8 AS col_61441,
			(e'\x0b/k\x17..':::STRING::STRING || tab_35645.col_61440::STRING[])::STRING[] AS col_61442
		FROM
			(VALUES (3531943256634900431:::INT8)) AS tab_35644 (col_61439),
			(
				VALUES
					(ARRAY[e'\x19"KMK;e.':::STRING,'':::STRING,e'\x16G+*F\x16\x1eD':::STRING]),
					(ARRAY['1':::STRING,e'\x187.\x1f3#':::STRING,e'\r%=':::STRING,e'\x1c\x01\x06e\x079':::STRING])
			)
				AS tab_35645 (col_61440)
		ORDER BY
			tab_35644.col_61439 DESC, tab_35644.col_61439 ASC
		LIMIT
			22:::INT8;
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=empty/setting=no-ddl.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-18162

--- a/pkg/sql/colexec/colbuilder/execplan.go
+++ b/pkg/sql/colexec/colbuilder/execplan.go
@@ -1996,16 +1996,18 @@ func planProjectionOperators(
 		}
 		leftExpr, rightExpr := t.TypedLeft(), t.TypedRight()
 		if t.Operator.Symbol == treebin.Concat {
-			// Concat requires special handling.
+			// Concat requires special handling since it has special rules when
+			// one of the arguments is an array or a string. We don't have
+			// native vectorized support for arrays yet, so we don't have to do
+			// anything extra for them, but we do need to handle the string
+			// case.
 			leftType, rightType := leftExpr.ResolvedType(), rightExpr.ResolvedType()
-			if leftType.Family() != rightType.Family() {
-				// When we have two different types, we perform the STRING
-				// concatenation.
+			if t.Op.ReturnType == types.String && leftType.Family() != rightType.Family() {
+				// This is a special case of the STRING concatenation - we have
+				// to plan a cast of the non-string type to a STRING.
 				if leftType.Family() == types.StringFamily {
-					// Need to cast the right expr to the STRING.
 					rightExpr = tree.NewTypedCastExpr(rightExpr, types.String)
 				} else if rightType.Family() == types.StringFamily {
-					// Need to cast the left expr to the STRING.
 					leftExpr = tree.NewTypedCastExpr(leftExpr, types.String)
 				} else {
 					// This is unexpected.

--- a/pkg/sql/logictest/testdata/logic_test/array
+++ b/pkg/sql/logictest/testdata/logic_test/array
@@ -2263,3 +2263,11 @@ SELECT indkey FROM pg_index WHERE indrelid = (SELECT oid FROM pg_class WHERE rel
 
 statement ok
 RESET distsql_workmem
+
+# Regression test for incorrect planning in the vectorized engine for
+# concatenating a string with a string array (#85295).
+query T
+SELECT ('foo'::STRING || col::STRING[])::STRING[] FROM (VALUES (ARRAY['bar':::STRING]), (ARRAY['baz':::STRING])) AS t (col);
+----
+{foo,bar}
+{foo,baz}

